[Jetpack Setup] Integrate the Connection API for Application Passwords

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13630


### Description
This PR integrates the Jetpack Connection API with the Jetpack Setup flow, the following limitations applies for now:
- The integration works only when using Application Passwords, the other flow ill be handled separately.
- We don't support magic link authentication yet, if used, it will use the WebView flow.
- When an older version of Jetpack is installed, the app will use the WebView flow, but it's broken as of this PR, I'll look into this in a separate PR. 

The connection is using the following flow:

```mermaid
flowchart TD
    A["Fetch /connection/data<br>(We already have this)"]
    A --> Result{Result}
    Result --> Error["Error<br>(We already handle this case)"]
    Result --> Success
    Success --> Infer{"Jetapck Status:<br>IsJetpackinstalled,<br>IsRegistered, IsCurrentUserConnected,<br>blogId"}
    Infer --> |IsCurrentUserConnected = true| Connected[Already connected]
    Infer --> | IsRegistered = null | IsInstalled{"IsJetpackinstalled?"}
    Infer --> IsRegistered{IsRegistered?}
    Register[Register: /connection/register]
    IsRegistered --> |"No<br>(Check notes)"| Register
    IsRegistered --> |Yes| Provision
    Register --> Provision[Provision: /remote_provision]
    Provision --> Connect[Connect WPCom: /jetpack-remote-connect-user]

    IsInstalled --> |Yes| WebView["Use WebView Connection"]
    IsInstalled --> |No| InferIsRegisterd["IsRegistered = connectionOwner == null"]
    InferIsRegisterd --> IsRegistered
```

### Steps to reproduce
Check the testing plan here pe5sF9-42S-p2, test the Application Passwords tests, just skip TC6 and TC7.

### Testing information
- Check the TCs and confirm the behavior is correct.

### The tests that have been performed
^

### Images/gif
https://github.com/user-attachments/assets/6928452c-fca4-42b3-a242-74bc45bd6b37


- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Refactor the JetpackStatus model

hichamboushaba committed:
Simplify logic

hichamboushaba committed:
Rename property for better clarity

hichamboushaba committed:
Pass the JetpackStatus to the activation screen

hichamboushaba committed:
First pass at handling the Jetpack connection at the repository level

hichamboushaba committed:
Handle native jetpack connection

hichamboushaba committed:
Skip fetching the Jetpack email when using native connection

hichamboushaba committed:
Improve code to remove warnings

hichamboushaba committed:
Add unit tests

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Remove the temporary wpComEmail property

hichamboushaba committed:
Remove usages of the temporary `JetpackStatus.Companion.invoke` constructor

hichamboushaba committed:
Handle the case of Jeptack CP

When we receive the connection data but Jetpack is not installed, the `isRegistered` could be missing, so we infer the registration status based on the `connectionOwner` property

JorgeMucientes reviewed (approved):
Really neat @hichamboushaba. All test cases work as expected and the code for the new API connection flow looks great and its easy to follow! 👏🏼 

Ship it!
